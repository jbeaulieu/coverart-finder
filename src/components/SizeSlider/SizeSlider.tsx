import { Box, Grid, Input, Slider, Typography } from "@mui/material";

export type Props = {
  size: number;
  setSize: (size: number) => void;
};

export const MIN_IMAGE_SIZE_PX = 50;
export const MAX_IMAGE_SIZE_PX = 1000;

const SizeSlider = (props: Props) => {
  const { size, setSize } = props;

  const handleSliderChange = (_event: Event, newValue: number) => {
    setSize(newValue);
  };

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSize(event.target.value === '' ? MIN_IMAGE_SIZE_PX : Number(event.target.value));
  };

  const handleBlur = () => {
    if (size < MIN_IMAGE_SIZE_PX) {
      setSize(MIN_IMAGE_SIZE_PX);
    } else if (size > MAX_IMAGE_SIZE_PX) {
      setSize(MAX_IMAGE_SIZE_PX);
    }
  };

  return (
    <Box sx={{ width: 450 }}>
      <Typography id="input-slider" gutterBottom>
        Download Size (px)
      </Typography>
      <Grid container spacing={2} sx={{ alignItems: 'center' }}>
        <Grid size="grow">
          <Slider
            value={size}
            min={MIN_IMAGE_SIZE_PX}
            max={MAX_IMAGE_SIZE_PX}
            step={50}
            onChange={handleSliderChange}
            aria-labelledby="input-slider"
          />
        </Grid>
        <Grid>
          <Input
            value={size}
            size="small"
            onChange={handleInputChange}
            onBlur={handleBlur}
            inputProps={{
              step: 50,
              min: MIN_IMAGE_SIZE_PX,
              max: MAX_IMAGE_SIZE_PX,
              type: 'number',
              'aria-labelledby': 'input-slider',
            }}
          />
        </Grid>
      </Grid>
    </Box>
  );
};

export default SizeSlider;
