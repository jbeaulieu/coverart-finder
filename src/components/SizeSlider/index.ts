import SizeSlider from './SizeSlider';

export default SizeSlider;
