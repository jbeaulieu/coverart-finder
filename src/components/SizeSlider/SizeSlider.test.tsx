import { fireEvent, render, screen } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import SizeSlider, { MAX_IMAGE_SIZE_PX, MIN_IMAGE_SIZE_PX } from './SizeSlider';
import type { Props } from './SizeSlider';

const setSize = vi.fn();
const validSizeValue = 500;
const invalidSizeValueTooSmall = 20;
const invalidSizeValueTooLarge = 5000;

const defaultProps: Props = {
  size: 100,
  setSize,
};

const renderSizeSlider = (props = defaultProps) => {
  return render(<SizeSlider {...props} />);
};

describe('Component SizeSlider', () => {

  describe('on click', () => {
    it('renders without crashing', () => {
      const { asFragment } = renderSizeSlider();

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('on slider update', () => {
    beforeEach(() => {
      setSize.mockClear();
    });

    it('invokes setSize function for valid values', () => {
      renderSizeSlider();

      const slider = screen.getByRole("slider", { name: "Download Size (px)" });
      fireEvent.change(slider, { target: { value: validSizeValue } });

      expect(setSize).toHaveBeenCalledExactlyOnceWith(validSizeValue);
    });

    it('invokes setSize function with MIN_IMAGE_SIZE_PX for values below the min size', () => {
      renderSizeSlider();

      const slider = screen.getByRole("slider", { name: "Download Size (px)" });
      fireEvent.change(slider, { target: { value: invalidSizeValueTooSmall } });

      expect(setSize).toHaveBeenCalledExactlyOnceWith(MIN_IMAGE_SIZE_PX);
    });

    it('invokes setSize function with MIN_IMAGE_SIZE_PX for values below the min size', () => {
      renderSizeSlider();

      const slider = screen.getByRole("slider", { name: "Download Size (px)" });
      fireEvent.change(slider, { target: { value: invalidSizeValueTooLarge } });

      expect(setSize).toHaveBeenCalledExactlyOnceWith(MAX_IMAGE_SIZE_PX);
    });
  });

  describe('on input field update', () => {
    beforeEach(() => {
      setSize.mockClear();
    });

    // Because the input field has increment/decrement buttons attached,
    // React assigns it a "spinbutton" role, rather than a generic "input"

    it('invokes setSize function for valid values', () => {
      renderSizeSlider();

      const input = screen.getByRole("spinbutton");
      fireEvent.change(input, { target: { value: validSizeValue } });

      expect(setSize).toHaveBeenCalledExactlyOnceWith(validSizeValue);
    });

    it('invokes setSize function with MIN_IMAGE_SIZE_PX for non-numerical values', () => {
      renderSizeSlider();

      const input = screen.getByRole("spinbutton");
      fireEvent.input(input, { target: { value: 'non-numerical value' } });

      expect(setSize).toHaveBeenCalledExactlyOnceWith(MIN_IMAGE_SIZE_PX);
    });
  });
});
