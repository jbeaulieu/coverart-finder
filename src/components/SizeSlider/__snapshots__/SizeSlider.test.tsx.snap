// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component SizeSlider > on click > renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-ve9ogo"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-vjrxpc-MuiTypography-root"
      id="input-slider"
    >
      Download Size (px)
    </p>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-mwb7rv-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-grow css-1ynczkt-MuiGrid-root"
      >
        <span
          class="MuiSlider-root MuiSlider-colorPrimary MuiSlider-sizeMedium css-13vpke-MuiSlider-root"
        >
          <span
            class="MuiSlider-rail css-r64h58-MuiSlider-rail"
          />
          <span
            class="MuiSlider-track css-xvk2i-MuiSlider-track"
            style="left: 0%; width: 5.2631578947368425%;"
          />
          <span
            class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-1xcmt9q-MuiSlider-thumb"
            data-index="0"
            style="left: 5.2631578947368425%;"
          >
            <input
              aria-labelledby="input-slider"
              aria-orientation="horizontal"
              aria-valuemax="1000"
              aria-valuemin="50"
              aria-valuenow="100"
              data-index="0"
              max="1000"
              min="50"
              step="50"
              style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
              type="range"
              value="100"
            />
          </span>
        </span>
      </div>
      <div
        class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-sizeSmall css-1cmet5c-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-labelledby="input-slider"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputSizeSmall css-1wd3yy0-MuiInputBase-input-MuiInput-input"
            max="1000"
            min="50"
            step="50"
            type="number"
            value="100"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
